synapse/federation/federation_client.py

```diff
@@ -403,9 +432,24 @@ async def get_pdu(
                 continue
 
         if signed_pdu:
-            self._get_pdu_cache[event_id] = signed_pdu
+            self._get_pdu_cache[signed_pdu.event_id] = signed_pdu
+
+            # Make sure to return a copy because downstream callers will use this
+            # event reference directly and change our original, pristine, untouched
+            # PDU. For example when people mark the event as an `outlier`
+            # (`event.internal_metadata.outlier = true`), we don't want that to
+            # propagate back into the cache.
+            #
+            # We could get away with only making a new copy of the event when
+            # pulling from cache but it's probably better to have good hygiene and
+            # not dirty the cache in the first place as well.
+            event_copy = make_event_from_dict(
+                signed_pdu.get_pdu_json(),
+                signed_pdu.room_version,
+                internal_metadata_dict=None,
+            )
```

Comment:
it would be nice to avoid duplicating this code. My way of looking at it is: whether we've recently added it, or just pulled it out, we have a reference to an event which is in the `get_pdu_cache`. In either case, we make a copy of it before returning it to the caller. Could we stick the whole `for` loop inside a condition so that we only do it if we do not find an entry in the cache?